bst.py

```diff
@@ -0,0 +1,169 @@
+                "\n".join(self._get_dot())
+            )
+        ))
+
+    def _get_dot(self):
+        """recursively prepare a dot graph entry for this node."""
+        if self.left is not None:
+            yield "\t%s -> %s;" % (self.data, self.left.data)
+            for i in self.left._get_dot():
+                yield i
+        elif self.right is not None:
+            r = random.randint(0, 1e9)
+            yield "\tnull%s [shape=point];" % r
+            yield "\t%s -> null%s;" % (self.data, r)
+        if self.right is not None:
+            yield "\t%s -> %s;" % (self.data, self.right.data)
+            for i in self.right._get_dot():
+                yield i
+        elif self.left is not None:
+            r = random.randint(0, 1e9)
+            yield "\tnull%s [shape=point];" % r
+            yield "\t%s -> null%s;" % (self.data, r)
+
+class Bst(object):
+    top = None
+    def __init__(self):
+        """Initialize a binary search tree"""
+        self.top = None
+        self.set = set()
+        self.depth = 0
+        self.dleft = 1
+        self.dright = 1
+        global top
+        top = self.top
+
+    def insert(self, val):
+        """Insert a Node into the binary tree and tracks the
+        depth of the right and left branches"""
+        if val not in self.set:
+            count = 1
+            if not self.top:
+                self.top = Node(val)
+            else:
+                current = self.top
+                while current:
+                    if val < current.data:
+                        if current.left is None:
+                            current.left = Node(val)
+                            break
+                        current = current.left
+                    else:
+                        if current.right is None:
+                            current.right = Node(val)
+                            break
+                        current = current.right
+                    count += 1
+                count += 1
+            if count > self.depth:
+                self.depth = count
+            if val < self.top.data and count > self.dleft:
+                self.dleft = count
+            elif val > self.dright and count > self.dright:
+                self.dright = count
+            self.set.add(val)
+
+    def contains(self, val):
+        """Return a boolean value if value currently exits in the tree"""
+        return val in self.set
+
+    def size(self):
+        """Return the current number of items in the tree"""
+        return len(self.set)
+
+    def depth(self):
+        """Return the depth of the longest branch"""
+        return self.depth
+
+    def balance(self):
+        """Determines if the right is deeper than the left branch"""
+        return self.dleft - self.dright
+
+    def in_order(self):
+        """Return a generator of the values using in order traversal"""
+        l = self.in_order_helper(self.top, [])
+        for num in l:
+            yield num
```

Comment:
strictly speaking, this is a generator.  But since it first materializes the entire tree as a list, it kind of misses the point of making the generator in the first place.